### ISSUE TITLE
Invertible ops

### ISSUE THREAD

**Issue by sebastienwood** [microsoft/DeepSpeed#229]
Hi !
Less of an issue than a question, in [this](https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/) research blog there is a mention of an invertible Softmax.
I searched an implementation to make sure I understood it, and the differences with a normal softmax, without success. 
Can anyone point me where to look at ?
Thanks !

**Comment by samyam** [comment]
@sebastienwood We are preparing the code for release very soon. Please keep an eye on https://www.deepspeed.ai/ for when it arrives.

The invertible Softmax can compute the backward without requiring to store the input to the softmax simply based on the output because it can derive the input directly from the output in some sense. A simplified example would be y=exp(x). You can compute x from y as x = log(y). We use this principle to reduce the memory requirement for the softmax allowing us to fit larger batch sizes.

**Comment by jeffra** [comment]
Hi @sebastienwood, thanks for your interest! We will be releasing our optimized transformer kernels _very_ soon. This will include the code for invertible softmax. Stay tuned! :) https://www.deepspeed.ai/news/2020/05/18/bert-record.html

**Comment by jeffra** [comment]
Hi @sebastienwood, I wanted to give you an update. We've now released all our transformer kernel source and updated our BERT example code. You can find out more details in our deep dive: https://www.deepspeed.ai/news/2020/05/27/fastest-bert-training.html

If you have further questions feel free to re-open this issue but I'll close it for now.

**Comment by sebastienwood** [comment]
Thank you very much ! :)
